File: dotnet/Barchart.BinarySerializer/Types/BinarySerializerInt.cs
#region Using Statements

using Barchart.BinarySerializer.Buffers;

#endregion

namespace Barchart.BinarySerializer.Types;

/// <summary>
///     Reads (and writes) int values to (and from) a binary data source.
/// </summary>
public class BinarySerializerInt : IBinaryTypeSerializer<int>
{
    #region Constants

    private const int ENCODED_LENGTH_IN_BYTES = sizeof(int);

    #endregion

    #region Methods

    /// <inheritdoc />
    public void Encode(IDataBufferWriter writer, int value)
    {
        writer.WriteBytes(BitConverter.GetBytes(value));
    }

    /// <inheritdoc />
    public int Decode(IDataBufferReader reader)
    {
        return BitConverter.ToInt32(reader.ReadBytes(ENCODED_LENGTH_IN_BYTES));
    }

    /// <inheritdoc />
    public bool GetEquals(int a, int b)
    {
        return a.Equals(b);
    }

    #endregion
}

File: typescript/src/types/binary-serializer-int.ts
import { DataReader } from "../buffers/data-reader.interface";
import { DataWriter } from "../buffers/data-writer.interface";
import { BinaryTypeSerializer } from "./binary-type-serializer.interface";
import { assertIntegerInRange } from './validation';

const MINIMUM_VALUE = -0x80000000;
const MAXIMUM_VALUE = 0x7FFFFFFF;

/**
 * Reads (and writes) int values to (and from) a binary data source.
 *
 * @public
 * @exported
 * @implements {BinaryTypeSerializer<number>}
 */
export class BinarySerializerInt implements BinaryTypeSerializer<number> {
    get sizeInBytes(): number {
        return 4;
    }

    encode(writer: DataWriter, value: number): void {
        assertIntegerInRange(value, MINIMUM_VALUE, MAXIMUM_VALUE);

        const buffer = new ArrayBuffer(this.sizeInBytes);
        const view = new DataView(buffer);

        view.setInt32(0, value, true);

        writer.writeBytes(new Uint8Array(buffer));
    }

    decode(reader: DataReader): number {
        const valueBytes = reader.readBytes(this.sizeInBytes);

        return new DataView(valueBytes.buffer, valueBytes.byteOffset, valueBytes.byteLength).getInt32(0, true);
    }

    getEquals(a: number, b: number): boolean {
        return a === b;
    }
}
